aws/rust-runtime/aws-inlineable/src/s3_express.rs

```diff
@@ -0,0 +1,788 @@
+            let response = client
+                .create_session()
+                .bucket(bucket_name)
+                .session_mode(crate::types::SessionMode::ReadWrite)
+                .send()
+                .await?;
+
+            response
+                .credentials
+                .ok_or("no session credentials in response".into())
+        }
+    }
+
+    #[derive(Default)]
+    pub(crate) struct Builder {
+        time_source: Option<SharedTimeSource>,
+        buffer_time: Option<Duration>,
+    }
+
+    impl Builder {
+        pub(crate) fn time_source(mut self, time_source: impl TimeSource + 'static) -> Self {
+            self.set_time_source(time_source.into_shared());
+            self
+        }
+        pub(crate) fn set_time_source(&mut self, time_source: SharedTimeSource) -> &mut Self {
+            self.time_source = Some(time_source.into_shared());
+            self
+        }
+        #[allow(dead_code)]
+        pub(crate) fn buffer_time(mut self, buffer_time: Duration) -> Self {
+            self.set_buffer_time(Some(buffer_time));
+            self
+        }
+        #[allow(dead_code)]
+        pub(crate) fn set_buffer_time(&mut self, buffer_time: Option<Duration>) -> &mut Self {
+            self.buffer_time = buffer_time;
+            self
+        }
+        pub(crate) fn build(self) -> DefaultS3ExpressIdentityProvider {
+            DefaultS3ExpressIdentityProvider {
+                cache: S3ExpressIdentityCache::new(
+                    DEFAULT_MAX_CACHE_CAPACITY,
+                    self.time_source.unwrap_or_default(),
+                    self.buffer_time.unwrap_or(DEFAULT_BUFFER_TIME),
+                ),
+            }
+        }
+    }
+
+    impl ResolveIdentity for DefaultS3ExpressIdentityProvider {
+        fn resolve_identity<'a>(
+            &'a self,
+            runtime_components: &'a RuntimeComponents,
+            config_bag: &'a ConfigBag,
+        ) -> IdentityFuture<'a> {
+            IdentityFuture::new(async move { self.identity(runtime_components, config_bag).await })
+        }
+
+        fn cache_location(&self) -> IdentityCacheLocation {
+            IdentityCacheLocation::IdentityResolver
+        }
+    }
+}
+
+/// Supporting code for S3 Express runtime plugin
+pub(crate) mod runtime_plugin {
+    use aws_runtime::auth::SigV4SessionTokenNameOverride;
+    use aws_sigv4::http_request::{SignatureLocation, SigningSettings};
+    use aws_smithy_runtime_api::{box_error::BoxError, client::runtime_plugin::RuntimePlugin};
+    use aws_smithy_types::config_bag::{ConfigBag, FrozenLayer, Layer};
+    use aws_types::os_shim_internal::Env;
+
+    mod env {
+        pub(super) const S3_DISABLE_EXPRESS_SESSION_AUTH: &str =
+            "AWS_S3_DISABLE_EXPRESS_SESSION_AUTH";
+    }
+
+    #[derive(Debug)]
+    pub(crate) struct S3ExpressRuntimePlugin {
+        config: FrozenLayer,
+    }
+
+    impl S3ExpressRuntimePlugin {
+        pub(crate) fn new(config: FrozenLayer) -> Self {
```

Comment:
I should have caught this earlier, but I think it's a simple fix. Passing a config here seems to run counter to the pattern of runtime plugins only contributing to config rather than taking and augmenting it. I think you should refactor this to take a boolean: `disable_s3_express_session_auth` instead of a whole config. This may allow you to get rid of the needless config clone in the `base_client_runtime_plugins` function.

Reply:
Updated in 60e6a89